fix(compiler-cli): `readConfiguration` existing options should override options in tsconfig

alan-agius4 opened the pull request:
**fix(compiler-cli): extend `angularCompilerOptions` in tsconfig from node**
TypeScript supports non rooted extends, we should do the same

https://github.com/microsoft/TypeScript/blob/b346f5764e4d500ebdeff7086e43690ea533a305/src/compiler/commandLineParser.ts#L2603-L2628

Closes: #36715


**refactor(bazel): use `readConfiguration` to read config file**
With this change we clean up the parsing of tsconfig files.



**fix(compiler-cli): `readConfiguration` existing options should override options in tsconfig**
At the moment, when passing an Angular Compiler option
in the `existingOptions` it doesn't override the defined in the TSConfig.

petebacondarwin commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -133,13 +133,21 @@ export function calcProjectFileAndBasePath(
+  const mergedConfig = {
+    enableIvy: true,
+    ...tsOptions,
+    ...(config.angularCompilerOptions ?? {}),
```
I don't believe the `?? {}` is needed here: [TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UDeBDA-ALigZ2AJwEsA7AcwBooAjTXQ0gXwG4AoAMwFciBjYAgeyJRWfPgAoAHllSZY9AJTJmUZVC4DcqgawIkMcREigosAcgEQTFAHQ3xVlBUqngAdz4mmzekA)

petebacondarwin commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -133,13 +133,21 @@ export function calcProjectFileAndBasePath(
+    ...existingOptions,
+  };
+
+  // Coerce to boolean as `enableIvy` can be `ngtsc|true|false` here.
+  mergedConfig.enableIvy = !!mergedConfig.enableIvy;
```
Nice refactoring

petebacondarwin commented on packages/compiler-cli/test/perform_compile_spec.ts:
```diff
@@ -77,4 +78,28 @@ describe('perform_compile', () => {
        expect(options.enableIvy).toBe(false);
      });
+
+  it('"existingOptions" should override options defined in tsconfig', () => {
```
```suggestion
  it('should override options defined in tsconfig with those defined in `existingOptions`', () => {
```

petebacondarwin reviewed (changes requested):
Ugh, I hate how convoluted this whole area of code is.
I appreciate that this is not the fault of this PR.
But I wonder if we ought to take this opportunity to clean it up rather than patching it after the fact, which is basically what we are doing in this PR.

As I understand it there are "TypeScript" configurations and "Angular" configurations.
We have to effectively parse the `tsconfig.json` twice with two different tools to extract each.
On top of this the caller can pass in an object of programmatically defined `existingOptions` which should override those that appear in the `tsconfig.json`. But since this `existingOptions` (despite the typing) can contain Angular options too, we have to do this overriding twice.

So I think that the `config` object should only contain Angular options and the `parsed.options` should only contain TypeScript options.

I believe that the call to `ts.parseJsonConfigFileContent()` should already override the TypeScript options with those from `existingOptions` but I think that this will not apply the Angular parts of `existingOptions` over the top of those in `config`.

So I think that the right thing to do is to fix the `config` object so that it has the Angular options overridden by those in `existingOptions` before passing it to `createNgCompilerOptions()`, which should not have to care about overriding.

Does that sound right?

alan-agius4 committed:
fix(compiler-cli): `readConfiguration` existing options should override options in tsconfig

At the moment, when passing an Angular Compiler option
in the `existingOptions` it doesn't override the defined in the TSConfig.

alan-agius4 committed:
refactor(bazel): use `readConfiguration` to read config file

With this change we clean up the parsing of tsconfig files.

petebacondarwin commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -44,81 +44,49 @@ export function runOneBuild(args: string[], inputs?: {[path: string]: string}):
+    return false;
   }
 
+  const allowedNgCompilerOptionsOverrides = [
```
Might as well make this a `Set()` and then you can do much faster check below.

alan-agius4 replied:
Indeed.

petebacondarwin commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -132,58 +133,74 @@ export function calcProjectFileAndBasePath(
-              ...baseConfig.angularCompilerOptions
-            };
+          let existingNgCompilerOptions: NgCompilerOptions;
+          if (parentOptions && config.angularCompilerOptions) {
```
`parentOptions` will always be truthy, since it is initialized to `{}` above.

So we can drop this `if` and just do 

```ts
const existingNgCompilerOptions = {...config.angularCompilerOptions, ...parentOptions};
```

petebacondarwin commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -132,58 +133,74 @@ export function calcProjectFileAndBasePath(
-                extendedConfigPath :
-                absoluteFrom(`${extendedConfigPath}.json`);
+          if (config.extends && typeof config.extends === 'string') {
+            const extendsValue: string = config.extends;
+            const basePath = host.dirname(configFile);
+            let extendedConfigPath: AbsoluteFsPath|undefined;
+
+            if (extendsValue.startsWith('.') || fs.isRooted(extendsValue)) {
+              extendedConfigPath = host.resolve(host.dirname(configFile), config.extends);
+              extendedConfigPath = host.extname(extendedConfigPath) ?
+                  extendedConfigPath :
+                  absoluteFrom(`${extendedConfigPath}.json`);
+            } else {
+              // Path isn't a rooted or relative path, resolve like a module.
+              const {
+                resolvedModule,
+              } =
+                  ts.nodeModuleNameResolver(
+                      extendsValue, configFile,
+                      {moduleResolution: ts.ModuleResolutionKind.NodeJs, resolveJsonModule: true},
+                      parseConfigHost);
+              if (resolvedModule) {
+                extendedConfigPath = absoluteFrom(resolvedModule.resolvedFileName);
+              }
+            }
```
I would move this block of code into its own function. Something like

```ts
getExtendedConfigPath(extendsValue: string): AbsoluteFsPath|null;
```

petebacondarwin reviewed (approved):
LGTM - just a couple of suggested refactorings.

alan-agius4 commented:
@petebacondarwin updated.

alan-agius4 committed:
fix(compiler-cli): extend `angularCompilerOptions` in tsconfig from node

TypeScript supports non rooted extends, we should do the same

https://github.com/microsoft/TypeScript/blob/b346f5764e4d500ebdeff7086e43690ea533a305/src/compiler/commandLineParser.ts#L2603-L2628

Closes: #36715

josephperrott reviewed (approved):
LGTM

Reviewed-For: bazel

JoostK commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -44,81 +44,49 @@ export function runOneBuild(args: string[], inputs?: {[path: string]: string}):
+  const compilerOpts: ng.AngularCompilerOptions = {
+    ...config['angularCompilerOptions'],
+    ...tsOptions,
+    ...overrides,
```
I didn't know you could spread the result of `Object.entries` into an object literal, so I tried it and it looks like it doesn't work?

```ts
{...[['a', 1]]}
// -> {0: ["a", 1]}

{...Object.fromEntries([['a', 1]])}
// {a: 1}
```

which would be concerning if no tests fail

JoostK replied:
Additionally, it looks like `overrides` is what used to be `userConfig.angularCompilerOptions`? It used to be that Bazel configuration (`angularCompilerOptions` before) was preferred over user configuration (`userConfig.angularCompilerOptions`), but now this is the other way around?

Notice also that the previous behavior was pretty susceptible to bugs for e.g. boolean flags: an explicit `false` value would get overridden by a user's `true` value, but not the other way around?!

alan-agius4 replied:
Good catch.

Previously the user config was preferred example, indeed there does appear to be a bug with boolean flags.

```
angularCompilerOptions['i18nOutFormat'] = angularCompilerOptions['i18nOutFormat'] ||	
          userConfig.angularCompilerOptions.i18nOutFormat;
```

JoostK commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -44,81 +44,49 @@ export function runOneBuild(args: string[], inputs?: {[path: string]: string}):
-          userConfig.angularCompilerOptions.createExternalSymbolFactoryReexports;
-    }
+  const {bazelOpts, options: tsOptions, files, config} = parsedOptions;
+  const {errors: userErrors, options} = ng.readConfiguration(project);
```
I'd use `options: userOptions` here as that should make it more clear how it differs from `config`.

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,6 +222,43 @@ export function readConfiguration(
+    configFile: string, extendsValue: string, host: ConfigurationHost,
+    fs: FileSystem): AbsoluteFsPath|null {
+  const parseConfigHost = {
+    useCaseSensitiveFileNames: true,
```
This appears to have been inherited from code in master, but it's strange to me that this isn't using `fs.isCaseSentive()`. Also one can wonder why `readDirectory` and `readFile` don't get routed through `fs`. I think it's fine to not touch this in this PR, just noting it here.

JoostK replied:
Also this may benefit from being extracted to a function (e.g. `createConfigHost`) as it's now declared twice.

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,6 +222,43 @@ export function readConfiguration(
+    readFile: ts.sys.readFile
+  };
+
+  let extendedConfigPath: AbsoluteFsPath|undefined;
```
In the compiler we tend use `null` with an explicit initializer (in this case at least, as not all code paths below assign a value to `extendedConfigPath`).

alan-agius4 replied:
Will change to adhere to the code style.

JoostK commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -227,6 +222,43 @@ export function readConfiguration(
+    }
+  }
+
+  if (extendedConfigPath && host.exists(extendedConfigPath)) {
```
This would become `extendedConfigPath !== null`.

alan-agius4 commented on packages/compiler-cli/src/perform_compile.ts:
```diff
@@ -174,16 +164,24 @@ export function readConfiguration(
         };
 
-    const {config, error} = readExtendedConfigFile(projectFile);
+    const parseConfigHost = {
```
Actually this is `ts.ParseConfigHost`

alan-agius4 committed:
fixup! refactor(bazel): use `readConfiguration` to read config file

alan-agius4 committed:
fixup! fix(compiler-cli): extend `angularCompilerOptions` in tsconfig from node

alan-agius4 committed:
fixup! refactor(bazel): use `readConfiguration` to read config file

alan-agius4 commented:
**Caretaker**: can you please run a TGP as the ngc-wrapped changes can be risky? Thanks.

AndrewKushnir commented:
[Presubmit](http://test/OCL:356664930:BASE:356664940:1612936109271:1be0667d).

AndrewKushnir commented:
@alan-agius4 I've ran a presubmit (with a subset of targets) and there are some build failures from the scripts (g3-specific) that use `createNgCompilerOptions` function (which was removed in this PR).

@kyliau could you please help look into the failures to see if g3 scripts need refactoring or the PR should be update to avoid removing the `createNgCompilerOptions` function?

alan-agius4 commented:
I can re-add that method although I prefer not to.

AndrewKushnir commented:
@alan-agius4 FYI I've started a [new presubmit](http://test/OCL:356665121:BASE:356878807:1613012117726:94b352ac) after @kyliau submitted an internal refactoring to avoid using `createNgCompilerOptions` function (thanks Keen!). Will keep this thread updated.

AndrewKushnir commented:
@alan-agius4 FYI presubmits were successful for the changes in this PR, adding the "merge" label back. Thank you.

kyliau commented:
thanks @AndrewKushnir !

alan-agius4 commented:
Thanks @AndrewKushnir & @kyliau for your help.
